core/trino-main/src/main/java/io/trino/operator/ExchangeOperator.java

```diff
@@ -197,6 +215,290 @@ public Page getOutput()
+                }
+                future = initializationFuture;
+            }
+            future.set(null);
+        }
+    }
+
+    private static class DirectExchangeDataSource
+            implements ExchangeDataSource
+    {
+        private final DirectExchangeClient directExchangeClient;
+
+        private DirectExchangeDataSource(DirectExchangeClient directExchangeClient)
+        {
+            this.directExchangeClient = requireNonNull(directExchangeClient, "directExchangeClient is null");
+        }
+
+        @Override
+        public SerializedPage pollPage()
+        {
+            return directExchangeClient.pollPage();
+        }
+
+        @Override
+        public boolean isFinished()
+        {
+            return directExchangeClient.isFinished();
+        }
+
+        @Override
+        public ListenableFuture<Void> isBlocked()
+        {
+            return directExchangeClient.isBlocked();
+        }
+
+        @Override
+        public void addSplit(RemoteSplit remoteSplit)
+        {
+            DirectExchangeInput taskInput = (DirectExchangeInput) remoteSplit.getExchangeInput();
+            directExchangeClient.addLocation(taskInput.getTaskId(), taskInput.getLocation());
+        }
+
+        @Override
+        public void noMoreSplits()
+        {
+            directExchangeClient.noMoreLocations();
+        }
+
+        @Override
+        public OperatorInfo getInfo()
+        {
+            return directExchangeClient.getStatus();
+        }
+
+        @Override
+        public void close()
+        {
+            directExchangeClient.close();
+        }
+    }
+
+    private static class ExternalExchangeDataSource
+            implements ExchangeDataSource
+    {
+        private final ExchangeSource exchangeSource;
+        private final LocalMemoryContext systemMemoryContext;
+
+        private ExternalExchangeDataSource(ExchangeSource exchangeSource, LocalMemoryContext systemMemoryContext)
+        {
+            this.exchangeSource = requireNonNull(exchangeSource, "exchangeSource is null");
+            this.systemMemoryContext = requireNonNull(systemMemoryContext, "systemMemoryContext is null");
+        }
+
+        @Override
+        public SerializedPage pollPage()
+        {
+            Slice data = exchangeSource.read();
+            systemMemoryContext.setBytes(exchangeSource.getSystemMemoryUsage());
+            if (data == null) {
+                return null;
+            }
+            // TODO: Avoid extra memory copy
+            return readSerializedPage(data.getInput());
+        }
+
+        @Override
+        public boolean isFinished()
+        {
+            return exchangeSource.isFinished();
+        }
+
+        @Override
+        public ListenableFuture<Void> isBlocked()
+        {
+            return asVoid(toListenableFuture(exchangeSource.isBlocked()));
+        }
+
+        @Override
+        public void addSplit(RemoteSplit remoteSplit)
+        {
+            // ignore
```

Comment:
Actually it is not possible, as due to the implementation details the same split is being delivered multiple times from different threads. Added a comment.

Reply:
Still you can have an assert here if you do 
```java
                else {
                    if (dataSource instanceof DirectExchangeDataSource) {
                        // adding splits makes sense only for DirectExchangeDataSource. Data source for ExternalExchangeDataSource is fixed on construction.
                        dataSource.addSplit(remoteSplit);
                    }
                }
``` 
in io.trino.operator.ExchangeOperator.LazyExchangeDataSource#addSplit

Checking if `ExchangeSourceHandle` passed in split match the ones used for construction is tricky :/

Reply:
I was thinking more about it, and it feels it might be safer to require `ExchangeSourceHandle` implement `equals` and `hashcode`. This way the flow can be similar to what currently there for direct exchanges, but instead of ignoring duplicate locations an assertion can be added to verify that the "locations" are always the same.